src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -573,45 +413,94 @@ private async ValueTask WriteAsyncChunked<TIOAdapter>(TIOAdapter writeAdapter, R
-
             byte[] rentedBuffer = ArrayPool<byte>.Shared.Rent(buffer.Length + FrameOverhead);
             byte[] outBuffer = rentedBuffer;
 
-            SecurityStatusPal status = EncryptData(buffer, ref outBuffer, out int encryptedBytes);
+            SecurityStatusPal status;
+            int encryptedBytes;
+            ValueTask t;
+            while (true)
+            {
+                status = EncryptData(buffer, ref outBuffer, out encryptedBytes);
+
+                // This should be rate when renegotiation happens exactly when we want to write.
+                if (status.ErrorCode == SecurityStatusPalErrorCode.TryAgain)
+                {
+                    TaskCompletionSource<bool>? waiter = _handshakeWaiter;
+                    if (waiter != null)
+                    {
+                        t = writeAdapter.LockAsync(waiter);
+                        // We finished synchronously waiting for renegotiation. Try it again.
+                        if (t.IsCompletedSuccessfully)
+                        {
+                            continue;
+                        }
+
+                        return CompleteLockAndWriteAsync(t, rentedBuffer);
+                    }
+
+                    continue;
+                    // We are on Async path now. We need to wait for the Lock as well as for the write when it is finished.
+                }
+
+                break;
+            }
 
             if (status.ErrorCode != SecurityStatusPalErrorCode.OK)
             {
-                // Re-handshake status is not supported.
                 ArrayPool<byte>.Shared.Return(rentedBuffer);
                 return new ValueTask(Task.FromException(ExceptionDispatchInfo.SetCurrentStackTrace(new IOException(SR.net_io_encrypt, SslStreamPal.GetException(status)))));
             }
 
-            ValueTask t = writeAdapter.WriteAsync(outBuffer, 0, encryptedBytes);
+            t = writeAdapter.WriteAsync(outBuffer, 0, encryptedBytes);
             if (t.IsCompletedSuccessfully)
             {
                 ArrayPool<byte>.Shared.Return(rentedBuffer);
-                FinishWrite();
                 return t;
             }
             else
             {
-                return CompleteAsync(t, rentedBuffer);
+                return CompleteWriteAsync(t, rentedBuffer);
             }
 
-            async ValueTask WaitForWriteIOSlot(TIOAdapter wAdapter, Task lockTask, ReadOnlyMemory<byte> buff)
+            async ValueTask CompleteLockAndWriteAsync(ValueTask t, byte[] bufferToReturn)
             {
-                await lockTask.ConfigureAwait(false);
-                await WriteSingleChunk(wAdapter, buff).ConfigureAwait(false);
+                SecurityStatusPal status;
+                int encryptedBytes;
+
+                try
+                {
+                    await t;
+                    while (true)
+                    {
+                        status = EncryptData(buffer, ref outBuffer, out encryptedBytes);
```

Comment:
This local async function is capturing data from the outer function.  I think if you look at the compiled IL, you'll find this is causing allocations even when the async local function is never invoked, e.g. note the display class allocation at the beginning of M here:
https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABATARgLABQGADAAQY4B0AQgK4Bm9MUAzgNyEnk4CsHBnAMzkspAMKkA3oVKzywjElIBZABQBKKTLk7qATwAuMANoBdUsAZMopALykAglCgBDPQAUIEADYAeYIYwAHyUAMoAFi6wACaUAEowAHYGqjjq/Do6GADspABiXhoZmbLaJRgAHORKBRAaZZnSBCXlAJzVlACaAJYw3tFFDS2Ozm6ePr76RiERUTCxCQa0UImqlozM6UNyAL4NewQ7QA
You should pass in any state you need explicitly (capturing anything other than `this` is likely to cause such behavior); if you want to avoid the possibility of this mistake, make it a normal instance method rather than a local, or if you want to pass in `this` as well, make it a static local function.

Reply:
I did not know that, thanks. Would such allocation be visible in VS memory profiler? e.g. is that something I could find eventually?

Reply:
> Would such allocation be visible in VS memory profiler? 

Yup.

Reply:
(note the concern here is specifically for async local functions. For other local functions, the compiler can pass referenced state in by ref)